server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -586,7 +583,7 @@ public void onFailure(Exception e) {
                 : "a started primary with non-pending operation term must be in primary mode " + this.shardRouting;
             shardStateUpdated.countDown();
         }
-        if (currentRouting != null && currentRouting.active() == false && newRouting.active()) {
```

Comment:
I think this is right and that currentRouting can never be null. But if so, we should go all-in on it and remove the null checks and add an assert currentRouting != null in the beginning of the method.

Reply:
Yea, looking through the code this seems like it can never be `null` indeed. I added the assertion and removed the null checks.